### ISSUE TITLE
IntentDialog type definition doesn't match code

### ISSUE THREAD

**Issue by ejsmith** [Microsoft/BotBuilder#367]
https://github.com/Microsoft/BotBuilder/blob/7299945ac9adffeaeb8f03d6b0b82c2bce9bd54e/Node/src/dialogs/IntentDialog.ts#L256
`protected abstract recognizeIntents(language: string, utterance: string, callback: (err: Error, intents?: IIntent[], entities?: IEntity[]) => void): void;`

https://github.com/Microsoft/BotBuilder/blob/master/Node/src/botbuilder.d.ts#L1398
`protected abstract recognizeIntents(session: Session, callback: (err: Error, intents?: IIntent[], entities?: IEntity[]) => void): void;`

Is this out of date? I am trying to build a custom intent dialog and this signature does not appear to match up with the code or with what is actually passed into the dialog.

**Comment by Stevenic** [comment]
I probably forgot to update the signature at one point as `botbuilder.d.ts` is a little decoupled from the actual code.

One thing I'd tell you though is that the `IntentDialog` class in its current form may be going away at some point. I was planning to just fold it into the `LuisDialog` class and make that close more purposeful.  As a starting point I'd probably recommend just making a copy of the IntentDailog class and use that as a template for your own custom dialog. You could also use `CommandDialog` as a starting point if you don't need the `IntentGroup` stuff.

Out of curiosity can I ask what you're building? If its something cool you should package it up as a library :) I understand if you can't share. Just curious more than anything.

**Comment by ejsmith** [comment]
Aren't the type definitions generated from the code since the code is written in TypeScript?

I wrote an IntentDialog for wit.ai and then another one for api.ai. So far I like api.ai better. Why are you thinking about getting rid of the IntentDialog?

I am building a slack bot, with a Justin Bieber personality :-) to do work tasks. I have a really nice bot app setup. Might pull our company specific code out and make it OSS.

**Comment by Stevenic** [comment]
No... I had to hand roll the .d.ts file in this case as the TypeScript compiler wasn't able to generate the right file. TS isn't quite there for NPM building yet.

Very cool on the wit.ai & api.ai support. Both are on our todo list to build as well.  LUIS is going in some very specific directions, as are wit.ai & api.ai.  These services have a fair amount in common today but doesn't look like that will be the case for long.  The IntentDialog abstraction is good because it lets you make all of these services look roughly the same but it results in lowest common denominator dialogs that can't easily take advantage of what's special about each service. After some reflection we've decided that we don't really want to promote lowest common denominator so it's probably better to implement a rich dialog specific to each service.  These dialogs can still follow similar patterns so that they feel familiar to the developer regardless of the intent service they pick.
